---a/net/minecraft/world/Teleporter.java
+++b/net/minecraft/world/Teleporter.java
@@ -1,10 +1,8 @@
 package net.minecraft.world;
 
-import com.google.common.collect.Lists;
 import java.util.Iterator;
 import java.util.List;
 import java.util.Random;
+
 import net.minecraft.block.BlockPortal;
 import net.minecraft.block.state.IBlockState;
 import net.minecraft.entity.Entity;
@@ -13,101 +13,155 @@
 import net.minecraft.util.LongHashMap;
 import net.minecraft.util.MathHelper;
 
-public class Teleporter
-{
+import org.bukkit.Location;
+import org.bukkit.event.entity.EntityPortalExitEvent;
+import org.bukkit.util.Vector;
+
+import com.google.common.collect.Lists;
+
+public class Teleporter {
+
     private final WorldServer worldServerInstance;
     private final Random random;
     private final LongHashMap destinationCoordinateCache = new LongHashMap();
     private final List destinationCoordinateKeys = Lists.newArrayList();
     private static final String __OBFID = "CL_00000153";
 
-    public Teleporter(WorldServer worldIn)
-    {
+    public Teleporter(WorldServer worldIn) {
         this.worldServerInstance = worldIn;
         this.random = new Random(worldIn.getSeed());
     }
 
-    public void placeInPortal(Entity entityIn, float rotationYaw)
-    {
-        if (this.worldServerInstance.provider.getDimensionId() != 1)
-        {
-            if (!this.placeInExistingPortal(entityIn, rotationYaw))
-            {
+    public void placeInPortal(Entity entityIn, float rotationYaw) {
+        if (this.worldServerInstance.provider.getDimensionId() != 1) {
+            if (!this.placeInExistingPortal(entityIn, rotationYaw)) {
                 this.makePortal(entityIn);
                 this.placeInExistingPortal(entityIn, rotationYaw);
             }
-        }
-        else
-        {
+        } else {
             int i = MathHelper.floor_double(entityIn.posX);
             int j = MathHelper.floor_double(entityIn.posY) - 1;
             int k = MathHelper.floor_double(entityIn.posZ);
-            byte b0 = 1;
-            byte b1 = 0;
+            // CraftBukkit start - Modularize end portal creation
+            BlockPos created = this.createEndPortal(entityIn.posX, entityIn.posY, entityIn.posZ);
+            entityIn.setLocationAndAngles((double) created.getX(), (double) created.getY(), (double) created.getZ(), entityIn.rotationYaw, 0.0F);
+            entityIn.motionX = entityIn.motionY = entityIn.motionZ = 0.0D;
+        }
+    }
 
-            for (int l = -2; l <= 2; ++l)
-            {
-                for (int i1 = -2; i1 <= 2; ++i1)
-                {
-                    for (int j1 = -1; j1 < 3; ++j1)
-                    {
-                        int k1 = i + i1 * b0 + l * b1;
-                        int l1 = j + j1;
-                        int i2 = k + i1 * b1 - l * b0;
-                        boolean flag = j1 < 0;
-                        this.worldServerInstance.setBlockState(new BlockPos(k1, l1, i2), flag ? Blocks.obsidian.getDefaultState() : Blocks.air.getDefaultState());
+    // Split out from original a(Entity, double, double, double, float) method in order to enable being called from createPortal
+    private BlockPos createEndPortal(double x, double y, double z) {
+        int i = MathHelper.floor_double(x);
+        int j = MathHelper.floor_double(y) - 1;
+        int k = MathHelper.floor_double(z);
+        byte b0 = 1;
+        byte b1 = 0;
+        // CraftBukkit end
+
+        for (int l = -2; l <= 2; ++l) {
+            for (int i1 = -2; i1 <= 2; ++i1) {
+                for (int j1 = -1; j1 < 3; ++j1) {
+                    int k1 = i + i1 * b0 + l * b1;
+                    int l1 = j + j1;
+                    int i2 = k + i1 * b1 - l * b0;
+                    boolean flag = j1 < 0;
+                    this.worldServerInstance.setBlockState(new BlockPos(k1, l1, i2), flag ? Blocks.obsidian.getDefaultState() : Blocks.air.getDefaultState());
+                }
+            }
+        }
+
+        // CraftBukkit start
+        return new BlockPos(i, k, k);
+    }
+
+    // use logic based on creation to verify end portal
+    private BlockPos findEndPortal(BlockPos portal) {
+        int i = portal.getX();
+        int j = portal.getY() - 1;
+        int k = portal.getZ();
+        byte b0 = 1;
+        byte b1 = 0;
+
+        for (int l = -2; l <= 2; ++l) {
+            for (int i1 = -2; i1 <= 2; ++i1) {
+                for (int j1 = -1; j1 < 3; ++j1) {
+                    int k1 = i + i1 * b0 + l * b1;
+                    int l1 = j + j1;
+                    int i2 = k + i1 * b1 - l * b0;
+                    boolean flag = j1 < 0;
+
+                    if (this.worldServerInstance.getBlockState(new BlockPos(k1, l1, i2)).getBlock() != (flag ? Blocks.obsidian : Blocks.air)) {
+                        return null;
                     }
                 }
             }
+        }
+        return new BlockPos(i, j, k);
+    }
+    // CraftBukkit end
 
-            entityIn.setLocationAndAngles((double)i, (double)j, (double)k, entityIn.rotationYaw, 0.0F);
-            entityIn.motionX = entityIn.motionY = entityIn.motionZ = 0.0D;
+    public boolean placeInExistingPortal(Entity entityIn, float p_180620_2_) {
+        // CraftBukkit start - Modularize portal search process and entity teleportation
+        BlockPos found = this.findPortal(entityIn.posX, entityIn.posY, entityIn.posZ, 128);
+        if (found == null) {
+            return false;
         }
+
+        Location exit = new Location(this.worldServerInstance.getWorld(), found.getX(), found.getY(), found.getZ(), p_180620_2_, entityIn.rotationPitch);
+        Vector velocity = entityIn.getBukkitEntity().getVelocity();
+        this.adjustExit(entityIn, exit, velocity);
+        entityIn.setLocationAndAngles(exit.getX(), exit.getY(), exit.getZ(), exit.getYaw(), exit.getPitch());
+        if (entityIn.motionX != velocity.getX() || entityIn.motionY != velocity.getY() || entityIn.motionZ != velocity.getZ()) {
+            entityIn.getBukkitEntity().setVelocity(velocity);
+        }
+        return true;
     }
 
-    public boolean placeInExistingPortal(Entity entityIn, float p_180620_2_)
-    {
-        boolean flag = true;
+    public BlockPos findPortal(double x, double y, double z, int short1) {
+        if (this.worldServerInstance.getWorld().getEnvironment() == org.bukkit.World.Environment.THE_END) {
+            return this.findEndPortal(this.worldServerInstance.provider.getSpawnCoordinate());
+        }
+        // CraftBukkit end
         double d0 = -1.0D;
-        int i = MathHelper.floor_double(entityIn.posX);
-        int j = MathHelper.floor_double(entityIn.posZ);
+        // CraftBukkit start
+        int i = MathHelper.floor_double(x);
+        int j = MathHelper.floor_double(z);
+        // CraftBukkit end
         boolean flag1 = true;
         Object object = BlockPos.ORIGIN;
         long k = ChunkCoordIntPair.chunkXZ2Int(i, j);
 
-        if (this.destinationCoordinateCache.containsItem(k))
-        {
-            Teleporter.PortalPosition portalposition = (Teleporter.PortalPosition)this.destinationCoordinateCache.getValueByKey(k);
+        if (this.destinationCoordinateCache.containsItem(k)) {
+            Teleporter.PortalPosition portalposition = (Teleporter.PortalPosition) this.destinationCoordinateCache.getValueByKey(k);
             d0 = 0.0D;
             object = portalposition;
             portalposition.lastUpdateTime = this.worldServerInstance.getTotalWorldTime();
             flag1 = false;
-        }
-        else
-        {
-            BlockPos blockpos4 = new BlockPos(entityIn);
+        } else {
+            BlockPos blockpos4 = new BlockPos(x, y, z); // CraftBukkit
 
-            for (int l = -128; l <= 128; ++l)
-            {
+            for (int l = -128; l <= 128; ++l) {
                 BlockPos blockpos1;
 
-                for (int i1 = -128; i1 <= 128; ++i1)
-                {
-                    for (BlockPos blockpos = blockpos4.add(l, this.worldServerInstance.getActualHeight() - 1 - blockpos4.getY(), i1); blockpos.getY() >= 0; blockpos = blockpos1)
-                    {
+                for (int i1 = -128; i1 <= 128; ++i1) {
+                    for (BlockPos blockpos = blockpos4.add(l, this.worldServerInstance.getActualHeight() - 1 - blockpos4.getY(), i1); blockpos.getY() >= 0; blockpos = blockpos1) {
                         blockpos1 = blockpos.down();
 
-                        if (this.worldServerInstance.getBlockState(blockpos).getBlock() == Blocks.portal)
-                        {
-                            while (this.worldServerInstance.getBlockState(blockpos1 = blockpos.down()).getBlock() == Blocks.portal)
-                            {
+                        if (this.worldServerInstance.getBlockState(blockpos).getBlock() == Blocks.portal) {
+                            while (this.worldServerInstance.getBlockState(blockpos1 = blockpos.down()).getBlock() == Blocks.portal) {
                                 blockpos = blockpos1;
                             }
 
                             double d1 = blockpos.distanceSq(blockpos4);
 
-                            if (d0 < 0.0D || d1 < d0)
-                            {
+                            if (d0 < 0.0D || d1 < d0) {
                                 d0 = d1;
                                 object = blockpos;
                             }
@@ -117,58 +117,68 @@
             }
         }
 
-        if (d0 >= 0.0D)
-        {
-            if (flag1)
-            {
-                this.destinationCoordinateCache.add(k, new Teleporter.PortalPosition((BlockPos)object, this.worldServerInstance.getTotalWorldTime()));
+        if (d0 >= 0.0D) {
+            if (flag1) {
+                this.destinationCoordinateCache.add(k, new Teleporter.PortalPosition((BlockPos) object, this.worldServerInstance.getTotalWorldTime()));
                 this.destinationCoordinateKeys.add(Long.valueOf(k));
             }
+            // CraftBukkit start - Move entity teleportation logic into exit
+            return (BlockPos) object;
+        } else {
+            return null;
+        }
+    }
 
-            double d4 = (double)((BlockPos)object).getX() + 0.5D;
-            double d5 = (double)((BlockPos)object).getY() + 0.5D;
-            double d6 = (double)((BlockPos)object).getZ() + 0.5D;
+    // Entity repositioning logic split out from original b method and combined with repositioning logic for The End from original a method
+    public void adjustExit(Entity entity, Location position, Vector velocity) {
+        Location from = position.clone();
+        Vector before = velocity.clone();
+        BlockPos object = new BlockPos(position.getBlockX(), position.getBlockY(), position.getBlockZ());
+        float f = position.getYaw();
+
+        if (this.worldServerInstance.getWorld().getEnvironment() == org.bukkit.World.Environment.THE_END) {
+            position.setPitch(0.0F);
+            velocity.setX(0);
+            velocity.setY(0);
+            velocity.setZ(0);
+        } else {
+            // CraftBukkit end
+            double d4 = (double) ((BlockPos) object).getX() + 0.5D;
+            double d5 = (double) ((BlockPos) object).getY() + 0.5D;
+            double d6 = (double) ((BlockPos) object).getZ() + 0.5D;
             EnumFacing enumfacing = null;
 
-            if (this.worldServerInstance.getBlockState(((BlockPos)object).west()).getBlock() == Blocks.portal)
-            {
+            if (this.worldServerInstance.getBlockState(((BlockPos) object).west()).getBlock() == Blocks.portal) {
                 enumfacing = EnumFacing.NORTH;
             }
 
-            if (this.worldServerInstance.getBlockState(((BlockPos)object).east()).getBlock() == Blocks.portal)
-            {
+            if (this.worldServerInstance.getBlockState(((BlockPos) object).east()).getBlock() == Blocks.portal) {
                 enumfacing = EnumFacing.SOUTH;
             }
 
-            if (this.worldServerInstance.getBlockState(((BlockPos)object).north()).getBlock() == Blocks.portal)
-            {
+            if (this.worldServerInstance.getBlockState(((BlockPos) object).north()).getBlock() == Blocks.portal) {
                 enumfacing = EnumFacing.EAST;
             }
 
-            if (this.worldServerInstance.getBlockState(((BlockPos)object).south()).getBlock() == Blocks.portal)
-            {
+            if (this.worldServerInstance.getBlockState(((BlockPos) object).south()).getBlock() == Blocks.portal) {
                 enumfacing = EnumFacing.WEST;
             }
 
-            EnumFacing enumfacing1 = EnumFacing.getHorizontal(entityIn.getTeleportDirection());
+            EnumFacing enumfacing1 = EnumFacing.getHorizontal(entity.getTeleportDirection());
 
-            if (enumfacing != null)
-            {
+            if (enumfacing != null) {
                 EnumFacing enumfacing2 = enumfacing.rotateYCCW();
-                BlockPos blockpos2 = ((BlockPos)object).offset(enumfacing);
+                BlockPos blockpos2 = ((BlockPos) object).offset(enumfacing);
                 boolean flag2 = this.func_180265_a(blockpos2);
                 boolean flag3 = this.func_180265_a(blockpos2.offset(enumfacing2));
 
-                if (flag3 && flag2)
-                {
-                    object = ((BlockPos)object).offset(enumfacing2);
+                if (flag3 && flag2) {
+                    object = ((BlockPos) object).offset(enumfacing2);
                     enumfacing = enumfacing.getOpposite();
                     enumfacing2 = enumfacing2.getOpposite();
-                    BlockPos blockpos3 = ((BlockPos)object).offset(enumfacing);
+                    BlockPos blockpos3 = ((BlockPos) object).offset(enumfacing);
                     flag2 = this.func_180265_a(blockpos3);
                     flag3 = this.func_180265_a(blockpos3.offset(enumfacing2));
                 }
@@ -172,81 +172,95 @@
                 float f6 = 0.5F;
                 float f1 = 0.5F;
 
-                if (!flag3 && flag2)
-                {
+                if (!flag3 && flag2) {
                     f6 = 1.0F;
-                }
-                else if (flag3 && !flag2)
-                {
+                } else if (flag3 && !flag2) {
                     f6 = 0.0F;
-                }
-                else if (flag3)
-                {
+                } else if (flag3) {
                     f1 = 0.0F;
                 }
 
-                d4 = (double)((BlockPos)object).getX() + 0.5D;
-                d5 = (double)((BlockPos)object).getY() + 0.5D;
-                d6 = (double)((BlockPos)object).getZ() + 0.5D;
-                d4 += (double)((float)enumfacing2.getFrontOffsetX() * f6 + (float)enumfacing.getFrontOffsetX() * f1);
-                d6 += (double)((float)enumfacing2.getFrontOffsetZ() * f6 + (float)enumfacing.getFrontOffsetZ() * f1);
+                d4 = (double) ((BlockPos) object).getX() + 0.5D;
+                d5 = (double) ((BlockPos) object).getY() + 0.5D;
+                d6 = (double) ((BlockPos) object).getZ() + 0.5D;
+                d4 += (double) ((float) enumfacing2.getFrontOffsetX() * f6 + (float) enumfacing.getFrontOffsetX() * f1);
+                d6 += (double) ((float) enumfacing2.getFrontOffsetZ() * f6 + (float) enumfacing.getFrontOffsetZ() * f1);
                 float f2 = 0.0F;
                 float f3 = 0.0F;
                 float f4 = 0.0F;
                 float f5 = 0.0F;
 
-                if (enumfacing == enumfacing1)
-                {
+                if (enumfacing == enumfacing1) {
                     f2 = 1.0F;
                     f3 = 1.0F;
-                }
-                else if (enumfacing == enumfacing1.getOpposite())
-                {
+                } else if (enumfacing == enumfacing1.getOpposite()) {
                     f2 = -1.0F;
                     f3 = -1.0F;
-                }
-                else if (enumfacing == enumfacing1.rotateY())
-                {
+                } else if (enumfacing == enumfacing1.rotateY()) {
                     f4 = 1.0F;
                     f5 = -1.0F;
-                }
-                else
-                {
+                } else {
                     f4 = -1.0F;
                     f5 = 1.0F;
                 }
 
-                double d2 = entityIn.motionX;
-                double d3 = entityIn.motionZ;
-                entityIn.motionX = d2 * (double)f2 + d3 * (double)f5;
-                entityIn.motionZ = d2 * (double)f4 + d3 * (double)f3;
-                entityIn.rotationYaw = p_180620_2_ - (float)(enumfacing1.getHorizontalIndex() * 90) + (float)(enumfacing.getHorizontalIndex() * 90);
+                // CraftBukkit start
+                double d2 = velocity.getX();
+                double d3 = velocity.getZ();
+                // CraftBukkit end
+                velocity.setX(d2 * (double) f3 + d3 * (double) f6);
+                velocity.setZ(d2 * (double) f5 + d3 * (double) f4);
+                f = f - (float) (enumfacing1.getHorizontalIndex() * 90) + (float) (enumfacing.getHorizontalIndex() * 90);
+            } else {
+                velocity.setX(0);
+                velocity.setY(0);
+                velocity.setZ(0);
             }
-            else
-            {
-                entityIn.motionX = entityIn.motionY = entityIn.motionZ = 0.0D;
-            }
 
-            entityIn.setLocationAndAngles(d4, d5, d6, entityIn.rotationYaw, entityIn.rotationPitch);
-            return true;
+            // entityIn.setLocationAndAngles(d4, d5, d6, entityIn.rotationYaw, entityIn.rotationPitch);
+            position.setX(d4);
+            position.setY(d5);
+            position.setZ(d6);
+            position.setYaw(f);
         }
-        else
-        {
-            return false;
+        EntityPortalExitEvent event = new EntityPortalExitEvent(entity.getBukkitEntity(), from, position, before, velocity);
+        this.worldServerInstance.getServer().getPluginManager().callEvent(event);
+        Location to = event.getTo();
+        if (event.isCancelled() || to == null || !entity.isEntityAlive()) {
+            position.setX(from.getX());
+            position.setY(from.getY());
+            position.setZ(from.getZ());
+            position.setYaw(from.getYaw());
+            position.setPitch(from.getPitch());
+            velocity.copy(before);
+        } else {
+            position.setX(to.getX());
+            position.setY(to.getY());
+            position.setZ(to.getZ());
+            position.setYaw(to.getYaw());
+            position.setPitch(to.getPitch());
+            velocity.copy(event.getAfter()); // event.getAfter() will never be null, as setAfter() will cause an NPE if null is passed in
         }
+        // CraftBukkit end
     }
 
-    private boolean func_180265_a(BlockPos p_180265_1_)
-    {
+    private boolean func_180265_a(BlockPos p_180265_1_) {
         return !this.worldServerInstance.isAirBlock(p_180265_1_) || !this.worldServerInstance.isAirBlock(p_180265_1_.up());
     }
 
-    public boolean makePortal(Entity p_85188_1_)
-    {
-        byte b0 = 16;
+    public boolean makePortal(Entity p_85188_1_) {
+        // CraftBukkit start - Allow for portal creation to be based on coordinates instead of entity
+        return this.createPortal(p_85188_1_.posX, p_85188_1_.posY, p_85188_1_.posZ, 16);
+    }
+
+    public boolean createPortal(double x, double y, double z, int b0) {
+        if (this.worldServerInstance.getWorld().getEnvironment() == org.bukkit.World.Environment.THE_END) {
+            createEndPortal(x, y, z);
+            return true;
+        }
+        // CraftBukkit end
         double d0 = -1.0D;
-        int i = MathHelper.floor_double(p_85188_1_.posX);
-        int j = MathHelper.floor_double(p_85188_1_.posY);
-        int k = MathHelper.floor_double(p_85188_1_.posZ);
+        // CraftBukkit start
+        int i = MathHelper.floor_double(x);
+        int j = MathHelper.floor_double(y);
+        int k = MathHelper.floor_double(z);
+        // CraftBukkit end
         int l = i;
         int i1 = j;
         int j1 = k;
@@ -269,59 +269,47 @@
         double d3;
         double d4;
 
-        for (i2 = i - b0; i2 <= i + b0; ++i2)
-        {
-            d1 = (double)i2 + 0.5D - p_85188_1_.posX;
+        for (i2 = i - b0; i2 <= i + b0; ++i2) {
+            d1 = (double) i2 + 0.5D - x; // CraftBukkit
 
-            for (k2 = k - b0; k2 <= k + b0; ++k2)
-            {
-                d2 = (double)k2 + 0.5D - p_85188_1_.posZ;
+            for (k2 = k - b0; k2 <= k + b0; ++k2) {
+                d2 = (double) k2 + 0.5D - z; // CraftBukkit
                 label271:
 
-                for (i3 = this.worldServerInstance.getActualHeight() - 1; i3 >= 0; --i3)
-                {
-                    if (this.worldServerInstance.isAirBlock(new BlockPos(i2, i3, k2)))
-                    {
-                        while (i3 > 0 && this.worldServerInstance.isAirBlock(new BlockPos(i2, i3 - 1, k2)))
-                        {
+                for (i3 = this.worldServerInstance.getActualHeight() - 1; i3 >= 0; --i3) {
+                    if (this.worldServerInstance.isAirBlock(new BlockPos(i2, i3, k2))) {
+                        while (i3 > 0 && this.worldServerInstance.isAirBlock(new BlockPos(i2, i3 - 1, k2))) {
                             --i3;
                         }
 
-                        for (j3 = l1; j3 < l1 + 4; ++j3)
-                        {
+                        for (j3 = l1; j3 < l1 + 4; ++j3) {
                             k3 = j3 % 2;
                             l3 = 1 - k3;
 
-                            if (j3 % 4 >= 2)
-                            {
+                            if (j3 % 4 >= 2) {
                                 k3 = -k3;
                                 l3 = -l3;
                             }
 
-                            for (i4 = 0; i4 < 3; ++i4)
-                            {
-                                for (j4 = 0; j4 < 4; ++j4)
-                                {
-                                    for (k4 = -1; k4 < 4; ++k4)
-                                    {
+                            for (i4 = 0; i4 < 3; ++i4) {
+                                for (j4 = 0; j4 < 4; ++j4) {
+                                    for (k4 = -1; k4 < 4; ++k4) {
                                         l4 = i2 + (j4 - 1) * k3 + i4 * l3;
                                         i5 = i3 + k4;
                                         int j5 = k2 + (j4 - 1) * l3 - i4 * k3;
 
-                                        if (k4 < 0 && !this.worldServerInstance.getBlockState(new BlockPos(l4, i5, j5)).getBlock().getMaterial().isSolid() || k4 >= 0 && !this.worldServerInstance.isAirBlock(new BlockPos(l4, i5, j5)))
-                                        {
+                                        if (k4 < 0 && !this.worldServerInstance.getBlockState(new BlockPos(l4, i5, j5)).getBlock().getMaterial().isSolid() || k4 >= 0 && !this.worldServerInstance.isAirBlock(new BlockPos(l4, i5, j5))) {
                                             continue label271;
                                         }
                                     }
                                 }
                             }
 
-                            d3 = (double)i3 + 0.5D - p_85188_1_.posY;
+                            d3 = (double) i3 + 0.5D - y; // CraftBukkit
                             d4 = d1 * d1 + d3 * d3 + d2 * d2;
 
-                            if (d0 < 0.0D || d4 < d0)
-                            {
+                            if (d0 < 0.0D || d4 < d0) {
                                 d0 = d4;
                                 l = i2;
                                 i1 = i3;
@@ -333,54 +333,42 @@
             }
         }
 
-        if (d0 < 0.0D)
-        {
-            for (i2 = i - b0; i2 <= i + b0; ++i2)
-            {
-                d1 = (double)i2 + 0.5D - p_85188_1_.posX;
+        if (d0 < 0.0D) {
+            for (i2 = i - b0; i2 <= i + b0; ++i2) {
+                d1 = (double) i2 + 0.5D - x; // CraftBukkit
 
-                for (k2 = k - b0; k2 <= k + b0; ++k2)
-                {
-                    d2 = (double)k2 + 0.5D - p_85188_1_.posZ;
+                for (k2 = k - b0; k2 <= k + b0; ++k2) {
+                    d2 = (double) k2 + 0.5D - z; // CraftBukkit
                     label219:
 
-                    for (i3 = this.worldServerInstance.getActualHeight() - 1; i3 >= 0; --i3)
-                    {
-                        if (this.worldServerInstance.isAirBlock(new BlockPos(i2, i3, k2)))
-                        {
-                            while (i3 > 0 && this.worldServerInstance.isAirBlock(new BlockPos(i2, i3 - 1, k2)))
-                            {
+                    for (i3 = this.worldServerInstance.getActualHeight() - 1; i3 >= 0; --i3) {
+                        if (this.worldServerInstance.isAirBlock(new BlockPos(i2, i3, k2))) {
+                            while (i3 > 0 && this.worldServerInstance.isAirBlock(new BlockPos(i2, i3 - 1, k2))) {
                                 --i3;
                             }
 
-                            for (j3 = l1; j3 < l1 + 2; ++j3)
-                            {
+                            for (j3 = l1; j3 < l1 + 2; ++j3) {
                                 k3 = j3 % 2;
                                 l3 = 1 - k3;
 
-                                for (i4 = 0; i4 < 4; ++i4)
-                                {
-                                    for (j4 = -1; j4 < 4; ++j4)
-                                    {
+                                for (i4 = 0; i4 < 4; ++i4) {
+                                    for (j4 = -1; j4 < 4; ++j4) {
                                         k4 = i2 + (i4 - 1) * k3;
                                         l4 = i3 + j4;
                                         i5 = k2 + (i4 - 1) * l3;
 
-                                        if (j4 < 0 && !this.worldServerInstance.getBlockState(new BlockPos(k4, l4, i5)).getBlock().getMaterial().isSolid() || j4 >= 0 && !this.worldServerInstance.isAirBlock(new BlockPos(k4, l4, i5)))
-                                        {
+                                        if (j4 < 0 && !this.worldServerInstance.getBlockState(new BlockPos(k4, l4, i5)).getBlock().getMaterial().isSolid() || j4 >= 0 && !this.worldServerInstance.isAirBlock(new BlockPos(k4, l4, i5))) {
                                             continue label219;
                                         }
                                     }
                                 }
 
-                                d3 = (double)i3 + 0.5D - p_85188_1_.posY;
+                                d3 = (double) i3 + 0.5D - y; // CraftBukkit
                                 d4 = d1 * d1 + d3 * d3 + d2 * d2;
 
-                                if (d0 < 0.0D || d4 < d0)
-                                {
+                                if (d0 < 0.0D || d4 < d0) {
                                     d0 = d4;
                                     l = i2;
                                     i1 = i3;
@@ -397,19 +397,16 @@
         int l5 = k1 % 2;
         int l2 = 1 - l5;
 
-        if (k1 % 4 >= 2)
-        {
+        if (k1 % 4 >= 2) {
             l5 = -l5;
             l2 = -l2;
         }
 
-        if (d0 < 0.0D)
-        {
+        if (d0 < 0.0D) {
             i1 = MathHelper.clamp_int(i1, 70, this.worldServerInstance.getActualHeight() - 10);
             j2 = i1;
 
-            for (i3 = -1; i3 <= 1; ++i3)
-            {
-                for (j3 = 1; j3 < 3; ++j3)
-                {
-                    for (k3 = -1; k3 < 3; ++k3)
-                    {
+            for (i3 = -1; i3 <= 1; ++i3) {
+                for (j3 = 1; j3 < 3; ++j3) {
+                    for (k3 = -1; k3 < 3; ++k3) {
                         l3 = k5 + (j3 - 1) * l5 + i3 * l2;
                         i4 = j2 + k3;
                         j4 = k2 + (j3 - 1) * l2 - i3 * l5;
@@ -426,12 +426,9 @@
 
         IBlockState iblockstate = Blocks.portal.getDefaultState().withProperty(BlockPortal.AXIS, l5 != 0 ? EnumFacing.Axis.X : EnumFacing.Axis.Z);
 
-        for (j3 = 0; j3 < 4; ++j3)
-        {
-            for (k3 = 0; k3 < 4; ++k3)
-            {
-                for (l3 = -1; l3 < 4; ++l3)
-                {
+        for (j3 = 0; j3 < 4; ++j3) {
+            for (k3 = 0; k3 < 4; ++k3) {
+                for (l3 = -1; l3 < 4; ++l3) {
                     i4 = k5 + (k3 - 1) * l5;
                     j4 = j2 + l3;
                     k4 = k2 + (k3 - 1) * l2;
@@ -440,10 +440,8 @@
                 }
             }
 
-            for (k3 = 0; k3 < 4; ++k3)
-            {
-                for (l3 = -1; l3 < 4; ++l3)
-                {
+            for (k3 = 0; k3 < 4; ++k3) {
+                for (l3 = -1; l3 < 4; ++l3) {
                     i4 = k5 + (k3 - 1) * l5;
                     j4 = j2 + l3;
                     k4 = k2 + (k3 - 1) * l2;
@@ -455,22 +455,17 @@
         return true;
     }
 
-    public void removeStalePortalLocations(long p_85189_1_)
-    {
-        if (p_85189_1_ % 100L == 0L)
-        {
+    public void removeStalePortalLocations(long p_85189_1_) {
+        if (p_85189_1_ % 100L == 0L) {
             Iterator iterator = this.destinationCoordinateKeys.iterator();
             long j = p_85189_1_ - 600L;
 
-            while (iterator.hasNext())
-            {
-                Long olong = (Long)iterator.next();
-                Teleporter.PortalPosition portalposition = (Teleporter.PortalPosition)this.destinationCoordinateCache.getValueByKey(olong.longValue());
+            while (iterator.hasNext()) {
+                Long olong = (Long) iterator.next();
+                Teleporter.PortalPosition portalposition = (Teleporter.PortalPosition) this.destinationCoordinateCache.getValueByKey(olong.longValue());
 
-                if (portalposition == null || portalposition.lastUpdateTime < j)
-                {
+                if (portalposition == null || portalposition.lastUpdateTime < j) {
                     iterator.remove();
                     this.destinationCoordinateCache.remove(olong.longValue());
                 }
@@ -476,16 +476,15 @@
         }
     }
 
-    public class PortalPosition extends BlockPos
-    {
+    public class PortalPosition extends BlockPos {
+
         public long lastUpdateTime;
         private static final String __OBFID = "CL_00000154";
 
-        public PortalPosition(BlockPos pos, long p_i45747_3_)
-        {
+        public PortalPosition(BlockPos pos, long p_i45747_3_) {
             super(pos.getX(), pos.getY(), pos.getZ());
             this.lastUpdateTime = p_i45747_3_;
         }
     }
-}
+}
